File: tests/test_layer.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <stdexcept>
#include <cmath>

#include "doctest.h"
#include <SHARPlib/layer.h>


TEST_CASE("Testing HeightLayer structs") {
    CHECK_THROWS_AS(
        sharp::HeightLayer layer2( 1000.0, 0.0 ), const std::range_error&
        );
}


TEST_CASE("Testing PressureLayer structs") {
    CHECK_THROWS_AS(
        sharp::PressureLayer layer2( 100.0, 1000.0 ), const std::range_error&
        );
}

TEST_CASE("Testing layer bounds checking and searching") {
    constexpr int N = 10;
    constexpr float pres[N] = {1000.0, 900.0, 800.0, 700.0, 600.0,
                               500.0,  400.0, 300.0, 200.0, 100.0};

    sharp::PressureLayer out_of_bounds_1 = {1000.0, 50.0};
    sharp::PressureLayer out_of_bounds_2 = {1100.0, 100.0};
    sharp::PressureLayer out_of_bounds_3 = {1100.0, 50.0};
    sharp::PressureLayer out_of_bounds_4 = {50.0, 25.0};
    sharp::PressureLayer out_of_bounds_5 = {1150.0, 1100.0};

	const sharp::LayerIndex idx_1 = sharp::get_layer_index(out_of_bounds_1, pres, N);
	const sharp::LayerIndex idx_2 = sharp::get_layer_index(out_of_bounds_2, pres, N);
	const sharp::LayerIndex idx_3 = sharp::get_layer_index(out_of_bounds_3, pres, N);
	const sharp::LayerIndex idx_4 = sharp::get_layer_index(out_of_bounds_4, pres, N);
	const sharp::LayerIndex idx_5 = sharp::get_layer_index(out_of_bounds_5, pres, N);

	CHECK(idx_1.kbot == 0);
	CHECK(idx_1.ktop == 9);
	CHECK(idx_2.kbot == 0);
	CHECK(idx_2.ktop == 9);
	CHECK(idx_3.kbot == 0);
	CHECK(idx_3.ktop == 9);
	CHECK(idx_4.kbot == 9);
	CHECK(idx_4.ktop == 9);
	CHECK(idx_5.kbot == 0);
	CHECK(idx_5.ktop == 0);
}

TEST_CASE("Testing layer_max over pressure layer") {

    float pres[10] = {1000.0, 900.0, 800.0, 700.0, 600.0, 
                      500.0, 400.0, 300.0, 200.0, 100.0};
    float data[10] = {0, 0, 0, 0, 10, 0, 0, 0, 0, 0};
    sharp::PressureLayer layer1(1000.0, 600.0); // max at top of layer
    sharp::PressureLayer layer2(1100.0, 50.0);  // out of bounds recovery
    sharp::PressureLayer layer3(1000.0, 850.0); // max is not in layer
    sharp::PressureLayer layer4(800.0, 650.0); // max is just above layer
    sharp::PressureLayer layer5(550.0, 400.0); // max is just below layer
    float pmax = -9999.0;

    CHECK(sharp::layer_max(layer1, pres, data, 10, &pmax) == 10.0);
    CHECK(pmax == 600.0);
    CHECK(sharp::layer_max(layer2, pres, data, 10, nullptr) == 10.0);
    CHECK(sharp::layer_max(layer3, pres, data, 10, nullptr) ==  0.0);
    CHECK(sharp::layer_max(layer4, pres, data, 10, &pmax) ==  
                                     doctest::Approx(4.80749));
    CHECK(pmax == 650.0);
    CHECK(sharp::layer_max(layer5, pres, data, 10, &pmax) ==  
                                     doctest::Approx(5.22758));
}


TEST_CASE("Testing layer_min over pressure layer") {

    float pres[10] = {1000.0, 900.0, 800.0, 700.0, 600.0, 
                      500.0, 400.0, 300.0, 200.0, 100.0};
    float data[10] = {0, 0, 0, 0, -10, 0, 0, 0, 0, 0};
    sharp::PressureLayer layer1(1000.0, 600.0); // min at top of layer
    sharp::PressureLayer layer2(1100.0, 50.0);  // out of bounds recovery
    sharp::PressureLayer layer3(1000.0, 850.0); // min is not in layer
    sharp::PressureLayer layer4(800.0, 650.0); // min is just above layer
    sharp::PressureLayer layer5(550.0, 400.0); // min is just below layer

    CHECK(sharp::layer_min(layer1, pres, data, 10) == -10.0);
    CHECK(sharp::layer_min(layer2, pres, data, 10) == -10.0);
    CHECK(sharp::layer_min(layer3, pres, data, 10) ==   0.0);
    CHECK(sharp::layer_min(layer4, pres, data, 10) ==  
                                    doctest::Approx(-4.80749));
    CHECK(sharp::layer_min(layer5, pres, data, 10) ==  
                                    doctest::Approx(-5.22758));
}


TEST_CASE("Testing layer_max over height layer") {

    float hght[10] = {1000.0, 2000.0, 3000.0, 4000.0, 5000.0, 
                      6000.0, 7000.0, 8000.0, 9000.0, 10000.0};
    float data[10] = {0, 0, 0, 0, 10, 0, 0, 0, 0, 0};
    sharp::HeightLayer layer1(1000.0, 6000.0); // max at top of layer
    sharp::HeightLayer layer2(-100.0, 20000.0);  // out of bounds recovery
    sharp::HeightLayer layer3(6000.0, 10000.0); // max is not in layer
    sharp::HeightLayer layer4(1000.0, 4750.0); // max is just above layer
    sharp::HeightLayer layer5(5250.0, 9000.0); // max is just below layer

    CHECK(sharp::layer_max(layer1, hght, data, 10, nullptr) == 10.0);
    CHECK(sharp::layer_max(layer2, hght, data, 10, nullptr) == 10.0);
    CHECK(sharp::layer_max(layer3, hght, data, 10, nullptr) ==  0.0);
    CHECK(sharp::layer_max(layer4, hght, data, 10, nullptr) ==  7.5);
    CHECK(sharp::layer_max(layer5, hght, data, 10, nullptr) ==  7.5);
}


TEST_CASE("Testing layer_min over height layer") {

    float hght[10] = {1000.0, 2000.0, 3000.0, 4000.0, 5000.0, 
                      6000.0, 7000.0, 8000.0, 9000.0, 10000.0};
    float data[10] = {0, 0, 0, 0, -10, 0, 0, 0, 0, 0};
    sharp::HeightLayer layer1(1000.0, 6000.0); // min at top of layer
    sharp::HeightLayer layer2(-100.0, 20000.0);  // out of bounds recovery
    sharp::HeightLayer layer3(6000.0, 10000.0); // min is not in layer
    sharp::HeightLayer layer4(1000.0, 4750.0); // min is just above layer
    sharp::HeightLayer layer5(5250.0, 9000.0); // min is just below layer

    CHECK(sharp::layer_min(layer1, hght, data, 10) == -10.0);
    CHECK(sharp::layer_min(layer2, hght, data, 10) == -10.0);
    CHECK(sharp::layer_min(layer3, hght, data, 10) ==  0.0);
    CHECK(sharp::layer_min(layer4, hght, data, 10) ==  -7.5);
    CHECK(sharp::layer_min(layer5, hght, data, 10) ==  -7.5);
}

TEST_CASE("Testing layer_mean over a pressure layer") {

	float pres[10] = { 1000.0, 900.0, 800.0, 700.0, 600.0,
		500.0, 400.0, 300.0, 200.0, 100.0 };

	float data[10] = { 0, 0, 0, 0, 10, 0, 0, 0, 0, 0 };

	sharp::PressureLayer layer1(1000.0, 600.0); // max at top of layer
	sharp::PressureLayer layer2(1100.0, 50.0);  // out of bounds recovery
	sharp::PressureLayer layer3(1000.0, 100.0);  // full layer 
	
	CHECK(sharp::layer_mean(layer1, pres, data, 10) == 1.25);
	CHECK(sharp::layer_mean(layer2, pres, data, 10) == doctest::Approx(1.1111));
	CHECK(sharp::layer_mean(layer3, pres, data, 10) == doctest::Approx(1.1111));

}

File: tests/python/test_layer.py
import pytest
import numpy as np

from nwsspc.sharp.calc import layer
from nwsspc.sharp.calc import constants


def test_height_layer_construction():
    lyr1 = layer.HeightLayer(0, 3000)
    assert (lyr1.bottom == 0 and lyr1.top == 3000)

    with pytest.raises(ValueError):
        layer.HeightLayer(3000, 0)

    with pytest.raises(ValueError):
        layer.HeightLayer(0, np.nan)

    with pytest.raises(ValueError):
        layer.HeightLayer(np.nan, 3000)

    with pytest.raises(ValueError):
        layer.HeightLayer(np.nan, np.nan)

    with pytest.raises(ValueError):
        layer.HeightLayer(np.inf, 0)

    with pytest.raises(ValueError):
        layer.HeightLayer(0, np.inf)

    with pytest.raises(ValueError):
        layer.HeightLayer(np.inf, np.inf)

    with pytest.raises(ValueError):
        layer.HeightLayer(np.inf, np.nan)


def test_pressure_layer_construction():
    lyr1 = layer.PressureLayer(100000, 10000)
    assert (lyr1.bottom == 100000 and lyr1.top == 10000)

    with pytest.raises(ValueError):
        layer.PressureLayer(10000, 100000)

    with pytest.raises(ValueError):
        layer.PressureLayer(0, np.nan)

    with pytest.raises(ValueError):
        layer.PressureLayer(np.nan, 3000)

    with pytest.raises(ValueError):
        layer.PressureLayer(np.nan, np.nan)

    with pytest.raises(ValueError):
        layer.PressureLayer(np.inf, 0)

    with pytest.raises(ValueError):
        layer.PressureLayer(0, np.inf)

    with pytest.raises(ValueError):
        layer.PressureLayer(np.inf, np.inf)

    with pytest.raises(ValueError):
        layer.PressureLayer(np.inf, np.nan)


def test_layer_conversion():
    pres = np.arange(10000, 110000, 10000)[::-1]
    hght = np.array([0.0, 500.0, 1500.0, 2500.0, 4000.0,
                    5500.0, 7500.0, 8500.0, 10500.0, 12500.0])

    hght_lyr = layer.HeightLayer(0.0, 3000.0)
    pres_lyr = layer.PressureLayer(100000.0, 75000.0)

    new_pres_lyr = layer.height_layer_to_pressure(hght_lyr, pres, hght)
    new_hght_lyr = layer.pressure_layer_to_height(pres_lyr, pres, hght)

    assert (new_pres_lyr.bottom == 100000.0)
    assert (new_hght_lyr.bottom == 0.0)
    assert (new_pres_lyr.top == pytest.approx(66666.7))
    assert (new_hght_lyr.top == pytest.approx(1983.3248))

    # Check out of bounds behavior
    h_oob1 = layer.HeightLayer(-100, 250.0)
    h_oob2 = layer.HeightLayer(11500.0, 14000.0)
    p_oob1 = layer.PressureLayer(115000.0, 90000.0)
    p_oob2 = layer.PressureLayer(15000.0, 5000.0)

    oob1 = layer.height_layer_to_pressure(h_oob1, pres, hght)
    oob2 = layer.height_layer_to_pressure(h_oob2, pres, hght)
    oob3 = layer.pressure_layer_to_height(p_oob1, pres, hght)
    oob4 = layer.pressure_layer_to_height(p_oob2, pres, hght)

    assert (oob1.bottom == constants.MISSING)
    assert (oob2.bottom == constants.MISSING)
    assert (oob3.bottom == constants.MISSING)
    assert (oob4.bottom == constants.MISSING)
    assert (oob1.top == constants.MISSING)
    assert (oob2.top == constants.MISSING)
    assert (oob3.top == constants.MISSING)
    assert (oob4.top == constants.MISSING)

    # Make sure the enforcement of 1D arrays works
    with pytest.raises(TypeError):
        layer.height_layer_to_pressure(hght_lyr, pres.reshape((5, 2)), hght)

    with pytest.raises(TypeError):
        layer.pressure_layer_to_height(pres_lyr, pres.reshape((5, 2)), hght)


def test_layer_bounds_check():
    pres = np.arange(10000, 110000, 10000)[::-1]

    p_oob1 = layer.PressureLayer(100000.0, 5000.0)
    p_oob2 = layer.PressureLayer(110000.0, 10000.0)
    p_oob3 = layer.PressureLayer(110000.0, 5000.0)
    p_oob4 = layer.PressureLayer(5000.0, 2500.0)
    p_oob5 = layer.PressureLayer(115000.0, 110000.0)

    idx1 = layer.get_layer_index(p_oob1, pres)
    idx2 = layer.get_layer_index(p_oob2, pres)
    idx3 = layer.get_layer_index(p_oob3, pres)
    idx4 = layer.get_layer_index(p_oob4, pres)
    idx5 = layer.get_layer_index(p_oob5, pres)

    assert (idx1.kbot == 0)
    assert (idx1.ktop == 9)
    assert (idx2.kbot == 0)
    assert (idx2.ktop == 9)
    assert (idx3.kbot == 0)
    assert (idx3.ktop == 9)
    assert (idx4.kbot == 9)
    assert (idx4.ktop == 9)
    assert (idx5.kbot == 0)
    assert (idx5.ktop == 0)


def test_pressure_layer_min():
    pres = np.arange(10000, 110000, 10000)[::-1]
    data = np.array([0, 0, 0, 0, -10, 0, 0, 0, 0, 0])

    lyr1 = layer.PressureLayer(100000.0, 60000.0)  # min at top of layer
    lyr2 = layer.PressureLayer(110000.0, 50000.0)  # out of bounds recovery
    lyr3 = layer.PressureLayer(100000.0, 85000.0)  # min is not in layer
    lyr4 = layer.PressureLayer(80000.0, 65000.0)  # min is just above layer
    lyr5 = layer.PressureLayer(55000.0, 40000.0)  # min is just below layer

    assert (layer.layer_min(lyr1, pres, data) == (-10.0, 60000.0))
    assert (layer.layer_min(lyr2, pres, data) == (-10.0, 60000.0))
    assert (layer.layer_min(lyr3, pres, data) == (0.0, 100000.0))
    assert (layer.layer_min(lyr4, pres, data) ==
            (pytest.approx(-4.807509899), 65000.0))
    assert (layer.layer_min(lyr5, pres, data) ==
            (pytest.approx(-5.22757434), 55000.0))


def test_pressure_layer_max():
    pres = np.arange(10000, 110000, 10000)[::-1]
    data = np.array([0, 0, 0, 0, 10, 0, 0, 0, 0, 0])

    lyr1 = layer.PressureLayer(100000.0, 60000.0)  # min at top of layer
    lyr2 = layer.PressureLayer(110000.0, 50000.0)  # out of bounds recovery
    lyr3 = layer.PressureLayer(100000.0, 85000.0)  # min is not in layer
    lyr4 = layer.PressureLayer(80000.0, 65000.0)  # min is just above layer
    lyr5 = layer.PressureLayer(55000.0, 40000.0)  # min is just below layer

    assert (layer.layer_max(lyr1, pres, data) == (10.0, 60000.0))
    assert (layer.layer_max(lyr2, pres, data) == (10.0, 60000.0))
    assert (layer.layer_max(lyr3, pres, data) == (0.0, 100000.0))
    assert (layer.layer_max(lyr4, pres, data) ==
            (pytest.approx(4.807509899), 65000.0))
    assert (layer.layer_max(lyr5, pres, data) ==
            (pytest.approx(5.22757434), 55000.0))


def test_height_layer_min():
    hght = np.arange(1000.0, 11000.0, 1000.0)
    data = np.array([0, 0, 0, 0, -10, 0, 0, 0, 0, 0])

    lyr1 = layer.HeightLayer(1000.0, 6000.0)  # max at top of layer
    lyr2 = layer.HeightLayer(-100.0, 20000.0)  # out of bounds recovery
    lyr3 = layer.HeightLayer(6000.0, 10000.0)  # max is not in layer
    lyr4 = layer.HeightLayer(1000.0, 4750.0)  # max is just above layer
    lyr5 = layer.HeightLayer(5250.0, 9000.0)  # max is just below layer

    assert (layer.layer_min(lyr1, hght, data) == (-10.0, 5000.0))
    assert (layer.layer_min(lyr2, hght, data) == (-10.0, 5000.0))
    assert (layer.layer_min(lyr3, hght, data) == (0.0, 6000.0))
    assert (layer.layer_min(lyr4, hght, data) == (-7.5, 4750.0))
    assert (layer.layer_min(lyr5, hght, data) == (-7.5, 5250.0))


def test_height_layer_max():
    hght = np.arange(1000.0, 11000.0, 1000.0)
    data = np.array([0, 0, 0, 0, 10, 0, 0, 0, 0, 0])

    lyr1 = layer.HeightLayer(1000.0, 6000.0)  # max at top of layer
    lyr2 = layer.HeightLayer(-100.0, 20000.0)  # out of bounds recovery
    lyr3 = layer.HeightLayer(6000.0, 10000.0)  # max is not in layer
    lyr4 = layer.HeightLayer(1000.0, 4750.0)  # max is just above layer
    lyr5 = layer.HeightLayer(5250.0, 9000.0)  # max is just below layer

    assert (layer.layer_max(lyr1, hght, data) == (10.0, 5000.0))
    assert (layer.layer_max(lyr2, hght, data) == (10.0, 5000.0))
    assert (layer.layer_max(lyr3, hght, data) == (0.0, 6000.0))
    assert (layer.layer_max(lyr4, hght, data) == (7.5, 4750.0))
    assert (layer.layer_max(lyr5, hght, data) == (7.5, 5250.0))


def test_pressure_layer_mean():
    pres = np.arange(10000, 110000, 10000)[::-1]
    data = np.array([0, 0, 0, 0, 10.0, 0, 0, 0, 0, 0])

    lyr1 = layer.PressureLayer(100000.0, 60000.0)  # max at top of layer
    lyr2 = layer.PressureLayer(110000.0, 5000.0)  # out of bounds recovery
    lyr3 = layer.PressureLayer(100000.0, 10000.0)  # full layer

    assert (layer.layer_mean(lyr1, pres, data) == 1.25)
    assert (layer.layer_mean(lyr2, pres, data) == pytest.approx(1.111111))
    assert (layer.layer_mean(lyr3, pres, data) == pytest.approx(1.111111))
